File: agent/pkg/collector/collector.go
package collector

import (
	"context"
	"fmt"
	"runtime"
	"time"

	"github.com/shirou/gopsutil/v3/cpu"
	"github.com/shirou/gopsutil/v3/disk"
	"github.com/shirou/gopsutil/v3/host"
	"github.com/shirou/gopsutil/v3/load"
	"github.com/shirou/gopsutil/v3/mem"
	"github.com/shirou/gopsutil/v3/net"
	"github.com/shirou/gopsutil/v3/process"
)

// Version is set via ldflags at build time
var Version = "0.1.0"

// SystemInfo 包含系统的各种信息
type SystemInfo struct {
	Timestamp   time.Time     `json:"timestamp"`
	Hostname    string        `json:"hostname"`
	Platform    string        `json:"platform"`
	OS          string        `json:"os"`
	Version     string        `json:"version"` // 操作系统版本
	CPUInfo     CPUInfo       `json:"cpu"`
	MemoryInfo  MemoryInfo    `json:"memory"`
	DiskInfo    []DiskInfo    `json:"disks"`
	NetworkInfo []NetworkInfo `json:"network"`
	LoadInfo    LoadInfo      `json:"load"`
	ProcessCount int           `json:"process_count"`
	TcpCount     int           `json:"tcp_count"`
	UdpCount     int           `json:"udp_count"`
	BootTime     time.Time     `json:"boot_time"`
	AgentVersion string        `json:"agent_version"`
}

// CPUInfo 包含CPU相关信息
type CPUInfo struct {
	Usage       float64 `json:"usage"`
	Cores       int     `json:"cores"`
	ModelName   string  `json:"model_name"`
	Arch        string  `json:"arch"`
	Temperature float64 `json:"temperature,omitempty"`
}

// MemoryInfo 包含内存相关信息
type MemoryInfo struct {
	Total     uint64  `json:"total"`
	Used      uint64  `json:"used"`
	Free      uint64  `json:"free"`
	UsageRate float64 `json:"usage_rate"`
}

// DiskInfo 包含磁盘相关信息
type DiskInfo struct {
	Device     string  `json:"device"`
	MountPoint string  `json:"mount_point"`
	Total      uint64  `json:"total"`
	Used       uint64  `json:"used"`
	Free       uint64  `json:"free"`
	UsageRate  float64 `json:"usage_rate"`
	FSType     string  `json:"fs_type"`
}

// NetworkInfo 包含网络相关信息
type NetworkInfo struct {
	Interface   string `json:"interface"`
	BytesSent   uint64 `json:"bytes_sent"`
	BytesRecv   uint64 `json:"bytes_recv"`
	PacketsSent uint64 `json:"packets_sent"`
	PacketsRecv uint64 `json:"packets_recv"`
}

// LoadInfo 包含系统负载信息
type LoadInfo struct {
	Load1  float64 `json:"load1"`
	Load5  float64 `json:"load5"`
	Load15 float64 `json:"load15"`
}

// Collector 定义数据收集器接口
type Collector interface {
	Collect(ctx context.Context) (*SystemInfo, error)
	SetSkipConn(skip bool)
	SetSkipProcs(skip bool)
}

// DefaultCollector 是默认的数据收集器实现
type DefaultCollector struct {
	skipConn  bool
	skipProcs bool
}

// NewCollector 创建一个新的数据收集器
func NewCollector() Collector {
	return &DefaultCollector{}
}

func (c *DefaultCollector) SetSkipConn(skip bool)  { c.skipConn = skip }
func (c *DefaultCollector) SetSkipProcs(skip bool) { c.skipProcs = skip }

// Collect 收集系统信息
func (c *DefaultCollector) Collect(ctx context.Context) (*SystemInfo, error) {
	info := &SystemInfo{
		Timestamp: time.Now(),
	}

	// 获取主机信息
	hostInfo, err := host.Info()
	if err != nil {
		return nil, fmt.Errorf("获取主机信息失败: %w", err)
	}
	info.Hostname = hostInfo.Hostname
	info.Platform = hostInfo.Platform
	info.OS = hostInfo.OS
	// 设置操作系统版本，格式化为更有意义的信息
	info.Version = fmt.Sprintf("%s %s (%s)", hostInfo.Platform, hostInfo.PlatformVersion, hostInfo.KernelVersion)

	// 获取CPU信息
	cpuPercent, err := cpu.Percent(time.Second, false)
	if err != nil {
		return nil, fmt.Errorf("获取CPU使用率失败: %w", err)
	}

	cpuInfo, err := cpu.Info()
	if err != nil {
		return nil, fmt.Errorf("获取CPU信息失败: %w", err)
	}

	var modelName string
	if len(cpuInfo) > 0 {
		modelName = cpuInfo[0].ModelName
	}

	info.CPUInfo = CPUInfo{
		Usage:     cpuPercent[0],
		Cores:     runtime.NumCPU(),
		ModelName: modelName,
		Arch:      runtime.GOARCH,
	}

	// 获取内存信息
	memInfo, err := mem.VirtualMemory()
	if err != nil {
		return nil, fmt.Errorf("获取内存信息失败: %w", err)
	}

	info.MemoryInfo = MemoryInfo{
		Total:     memInfo.Total,
		Used:      memInfo.Used,
		Free:      memInfo.Free,
		UsageRate: memInfo.UsedPercent,
	}

	// 获取磁盘信息
	// 用 all=true 枚举，跳过伪/虚拟文件系统并按设备去重。
	// 容器（overlay 根、设备名非 /dev/*）用 all=false 会被过滤掉导致 0/0，
	// 因此最后兜底直接测量根目录 "/"。
	seenDisk := make(map[string]bool)
	rootCovered := false
	if partitions, err := disk.Partitions(true); err == nil {
		for _, partition := range partitions {
			if isVirtualFS(partition.Fstype) {
				continue
			}
			if partition.Device != "" && seenDisk[partition.Device] {
				continue
			}
			usage, err := disk.Usage(partition.Mountpoint)
			if err != nil || usage.Total == 0 {
				continue
			}
			if partition.Device != "" {
				seenDisk[partition.Device] = true
			}
			if partition.Mountpoint == "/" {
				rootCovered = true
			}
			info.DiskInfo = append(info.DiskInfo, DiskInfo{
				Device:     partition.Device,
				MountPoint: partition.Mountpoint,
				Total:      usage.Total,
				Used:       usage.Used,
				Free:       usage.Free,
				UsageRate:  usage.UsedPercent,
				FSType:     partition.Fstype,
			})
		}
	}
	// 兜底：若没有分区覆盖到根目录（典型：容器 overlay 根被跳过），直接测量 "/"
	if !rootCovered {
		if usage, err := disk.Usage("/"); err == nil && usage.Total > 0 {
			info.DiskInfo = append(info.DiskInfo, DiskInfo{
				Device:     "overlay",
				MountPoint: "/",
				Total:      usage.Total,
				Used:       usage.Used,
				Free:       usage.Free,
				UsageRate:  usage.UsedPercent,
				FSType:     usage.Fstype,
			})
		}
	}

	// 获取网络信息（跳过回环 lo，避免把内部流量计入总流量）
	netIOCounters, err := net.IOCounters(true)
	if err != nil {
		return nil, fmt.Errorf("获取网络信息失败: %w", err)
	}

	for _, netIO := range netIOCounters {
		if netIO.Name == "lo" {
			continue
		}
		networkInfo := NetworkInfo{
			Interface:   netIO.Name,
			BytesSent:   netIO.BytesSent,
			BytesRecv:   netIO.BytesRecv,
			PacketsSent: netIO.PacketsSent,
			PacketsRecv: netIO.PacketsRecv,
		}
		info.NetworkInfo = append(info.NetworkInfo, networkInfo)
	}

	// 获取系统负载
	loadAvg, err := load.Avg()
	if err == nil {
		info.LoadInfo = LoadInfo{
			Load1:  loadAvg.Load1,
			Load5:  loadAvg.Load5,
			Load15: loadAvg.Load15,
		}
	}

	// 进程数量
	if !c.skipProcs {
		pids, err := process.Pids()
		if err == nil {
			info.ProcessCount = len(pids)
		}
	}

	// TCP / UDP 连接数
	if !c.skipConn {
		conns, err := net.Connections("all")
		if err == nil {
			for _, conn := range conns {
				switch conn.Type {
				case 1: // TCP
					info.TcpCount++
				case 2: // UDP
					info.UdpCount++
				}
			}
		}
	}

	// 获取系统启动时间
	bootTime, err := host.BootTime()
	if err == nil {
		info.BootTime = time.Unix(int64(bootTime), 0)
	}

	// 设置 Agent 版本（从编译时 ldflags 注入）
	info.AgentVersion = Version

	return info, nil
}

// isVirtualFS 判断是否为伪/虚拟文件系统（不计入磁盘统计）。
// 注意：容器根 overlay 在此被跳过，由 disk.Usage("/") 兜底测量，
// 既能避免宿主机上重复计算 overlay 挂载，又能保证容器内能取到磁盘。
func isVirtualFS(fstype string) bool {
	switch fstype {
	case "tmpfs", "devtmpfs", "devfs", "overlay", "overlayfs", "aufs",
		"proc", "sysfs", "cgroup", "cgroup2", "pstore", "bpf", "tracefs",
		"debugfs", "securityfs", "configfs", "fusectl", "mqueue", "hugetlbfs",
		"ramfs", "nsfs", "autofs", "binfmt_misc", "squashfs", "fuse.lxcfs",
		"rpc_pipefs", "selinuxfs", "efivarfs", "none", "":
		return true
	}
	return false
}

File: agent-java/src/main/java/com/qltz/agent/collector/Collector.java
package com.qltz.agent.collector;

import oshi.SystemInfo;
import oshi.hardware.CentralProcessor;
import oshi.hardware.GlobalMemory;
import oshi.hardware.HWDiskStore;
import oshi.hardware.NetworkIF;
import oshi.software.os.OperatingSystem;

import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.SocketException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;

/**
 * 系统信息采集器，使用 OSHI 库进行跨平台采集。
 * 对应 Go 版本的 collector.DefaultCollector。
 */
public class Collector {

    public static final String VERSION = "0.1.0";

    private final SystemInfo si;
    private boolean skipConn;
    private boolean skipProcs;

    public Collector() {
        this.si = new SystemInfo();
    }

    public void setSkipConn(boolean skip) { this.skipConn = skip; }
    public void setSkipProcs(boolean skip) { this.skipProcs = skip; }

    /**
     * 采集系统信息
     */
    public SystemInfo collect() {
        SystemInfo info = new SystemInfo();
        info.setTimestamp(Instant.now());

        oshi.SystemInfo oshiSi = new oshi.SystemInfo();
        OperatingSystem os = oshiSi.getOperatingSystem();
        CentralProcessor cpu = oshiSi.getHardware().getProcessor();
        GlobalMemory mem = oshiSi.getHardware().getMemory();

        // 主机信息
        info.setHostname(getHostname());
        info.setOs(os.getFamily());
        info.setPlatform(System.getProperty("os.name"));
        info.setVersion(os.getVersionInfo() != null
                ? os.getFamily() + " " + os.getVersionInfo().getVersion() + " (" + os.getVersionInfo().getBuildNumber() + ")"
                : System.getProperty("os.version"));

        // CPU
        SystemInfo.CPUInfo cpuInfo = new SystemInfo.CPUInfo();
        cpuInfo.setCores(Runtime.getRuntime().availableProcessors());
        cpuInfo.setModelName(cpu.getProcessorIdentifier().getName());
        cpuInfo.setArch(System.getProperty("os.arch"));
        // CPU 使用率：使用 OSHI 计算（需要两次采样取差值，这里取一个较短间隔）
        long[] prevTicks = cpu.getSystemCpuLoadTicks();
        try { Thread.sleep(1000); } catch (InterruptedException ignored) {}
        double cpuLoad = cpu.getSystemCpuLoadBetweenTicks(prevTicks) * 100.0;
        cpuInfo.setUsage(Math.max(0, Math.min(100, cpuLoad)));
        // 温度（如果有传感器）
        try {
            double temp = oshiSi.getHardware().getSensors().getCpuTemperature();
            if (temp > 0) cpuInfo.setTemperature(temp);
        } catch (Exception ignored) {}
        info.setCpuInfo(cpuInfo);

        // 内存
        SystemInfo.MemoryInfo memInfo = new SystemInfo.MemoryInfo();
        memInfo.setTotal(mem.getTotal());
        memInfo.setUsed(mem.getTotal() - mem.getAvailable());
        memInfo.setFree(mem.getAvailable());
        if (mem.getTotal() > 0) {
            memInfo.setUsageRate(((double)(mem.getTotal() - mem.getAvailable()) / mem.getTotal()) * 100.0);
        }
        info.setMemoryInfo(memInfo);

        // 磁盘
        List<SystemInfo.DiskInfo> disks = new ArrayList<>();
        for (HWDiskStore disk : oshiSi.getHardware().getDiskStores()) {
            for (oshi.hardware.HWPartition part : disk.getPartitions()) {
                if (isVirtualFS(part.getType())) continue;
                SystemInfo.DiskInfo di = new SystemInfo.DiskInfo();
                di.setDevice(disk.getName());
                di.setMountPoint(part.getMountPoint());
                di.setTotal(part.getSize());
                // HWPartition doesn't directly give used/free; use FileStore as fallback
                try {
                    java.nio.file.FileStore fs = java.nio.file.Paths.get(part.getMountPoint()).getFileSystem().getFileStores()
                            .iterator().hasNext() ? null : null;
                    java.io.File f = new java.io.File(part.getMountPoint());
                    long total = f.getTotalSpace();
                    long free = f.getFreeSpace();
                    long used = total - free;
                    di.setTotal(total);
                    di.setUsed(used);
                    di.setFree(free);
                    di.setUsageRate(total > 0 ? ((double) used / total) * 100.0 : 0);
                } catch (Exception e) {
                    di.setTotal(0);
                    di.setUsed(0);
                    di.setFree(0);
                    di.setUsageRate(0);
                }
                di.setFsType(part.getType());
                disks.add(di);
            }
        }
        // 兜底：如果没有分区，测量根目录
        if (disks.isEmpty()) {
            try {
                java.io.File root = new java.io.File("/");
                SystemInfo.DiskInfo di = new SystemInfo.DiskInfo();
                di.setDevice("overlay");
                di.setMountPoint("/");
                di.setTotal(root.getTotalSpace());
                di.setFree(root.getFreeSpace());
                di.setUsed(root.getTotalSpace() - root.getFreeSpace());
                di.setUsageRate(root.getTotalSpace() > 0
                        ? ((double)(root.getTotalSpace() - root.getFreeSpace()) / root.getTotalSpace()) * 100.0 : 0);
                di.setFsType("overlay");
                disks.add(di);
            } catch (Exception ignored) {}
        }
        info.setDiskInfo(disks);

        // 网络
        List<SystemInfo.NetworkInfo> nets = new ArrayList<>();
        for (NetworkIF netIF : oshiSi.getHardware().getNetworkIFs()) {
            if ("lo".equals(netIF.getName())) continue;
            netIF.updateAttributes();
            SystemInfo.NetworkInfo ni = new SystemInfo.NetworkInfo();
            ni.setIface(netIF.getName());
            ni.setBytesSent(netIF.getBytesSent());
            ni.setBytesRecv(netIF.getBytesRecv());
            ni.setPacketsSent(netIF.getPacketsSent());
            ni.setPacketsRecv(netIF.getPacketsRecv());
            nets.add(ni);
        }
        info.setNetworkInfo(nets);

        // 系统负载
        SystemInfo.LoadInfo loadInfo = new SystemInfo.LoadInfo();
        double[] loadAvg = cpu.getSystemLoadAverage(3);
        if (loadAvg.length >= 3) {
            loadInfo.setLoad1(loadAvg[0]);
            loadInfo.setLoad5(loadAvg[1]);
            loadInfo.setLoad15(loadAvg[2]);
        }
        info.setLoadInfo(loadInfo);

        // 进程数
        if (!skipProcs) {
            info.setProcessCount(os.getProcessCount());
        }

        // TCP/UDP 连接数 (OSHI doesn't provide connection counts directly)
        if (!skipConn) {
            // On Linux/macOS we could parse /proc/net or use lsof, but OSHI doesn't expose this.
            // We leave it as 0; it's better than halting the collection.
        }

        // 启动时间
        info.setBootTime(Instant.ofEpochSecond(os.getSystemBootTime()));

        // Agent 版本
        info.setAgentVersion(VERSION);

        return info;
    }

    private String getHostname() {
        try {
            return InetAddress.getLocalHost().getHostName();
        } catch (Exception e) {
            return "unknown";
        }
    }

    /**
     * 获取本机第一个非回环 IPv4 地址
     */
    public static String getLocalIP() {
        try {
            Enumeration<NetworkInterface> ifaces = NetworkInterface.getNetworkInterfaces();
            while (ifaces.hasMoreElements()) {
                NetworkInterface iface = ifaces.nextElement();
                if (iface.isLoopback() || !iface.isUp()) continue;
                Enumeration<InetAddress> addrs = iface.getInetAddresses();
                while (addrs.hasMoreElements()) {
                    InetAddress addr = addrs.nextElement();
                    if (addr instanceof java.net.Inet4Address) {
                        return addr.getHostAddress();
                    }
                }
            }
        } catch (SocketException ignored) {}
        return "unknown";
    }

    /**
     * 判断是否为虚拟文件系统，对应 Go 的 isVirtualFS
     */
    private boolean isVirtualFS(String fstype) {
        if (fstype == null) return true;
        return switch (fstype.toLowerCase()) {
            case "tmpfs", "devtmpfs", "devfs", "overlay", "overlayfs", "aufs",
                 "proc", "sysfs", "cgroup", "cgroup2", "pstore", "bpf", "tracefs",
                 "debugfs", "securityfs", "configfs", "fusectl", "mqueue", "hugetlbfs",
                 "ramfs", "nsfs", "autofs", "binfmt_misc", "squashfs", "fuse.lxcfs",
                 "rpc_pipefs", "selinuxfs", "efivarfs", "none", "" -> true;
            default -> false;
        };
    }
}
